analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/ObjectConstraint.cs

```diff
@@ -33,6 +39,9 @@ internal sealed class ObjectConstraint : SymbolicConstraint
         protected override string Name =>
             this == Null ? nameof(Null) : nameof(NotNull);
 
-        private ObjectConstraint() { }
+        private ObjectConstraint(ObjectConstraintKind kind) =>
+            Kind = kind;
+
+        public ObjectConstraintKind Kind { get; }
```

Comment:
The enum comes in handy for pattern matching. This is also what F# generates for its discriminated unions (The constraints are DUs) and what Roslyn does with SyntaxNode. So it seems a reasonable thing to add.

Reply:
I like the `ConstraintKind`. Please do it in another PR. Because it will affect also all constraint from the old engine.

For example `StringConstraint` is a pain - it maintains bunch of members that we don't need at all and we will not migrate them to the new engine.

We might add some kind of `SonarSymbolicConstraint` with kind `Unknown` to close the gap. (I know the other way around is cleaner OOP design, but I want `SymbolicConstraint` to be the one for Roslyn and to have the Sonar-hell easily removable by a clean cut)

Reply:
I reverted the changes nd will create a PR next.